compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/transformers/body/resolve/FirExpressionsResolveTransformer.kt

```diff
@@ -187,9 +188,46 @@ class FirExpressionsResolveTransformer(transformer: FirBodyResolveTransformer) :
             }
 
         dataFlowAnalyzer.exitFunctionCall(completeInference, callCompleted)
+        if (callCompleted && completeInference.isArrayOfCall) {
+            return completeInference.toArrayOfCall().compose()
+        }
         return completeInference.compose()
     }
 
+    private val FirFunctionCall.isArrayOfCall: Boolean
+        get() {
+            val function: FirCallableDeclaration<*> = getOriginalFunction() ?: return false
+            return function is FirSimpleFunction &&
+                    // TODO: extend it to other variants, like emptyArray, intArrayOf, floatArrayOf, etc.
+                    !function.name.isSpecial && function.name.identifier == "arrayOf" &&
```

Comment:
use name.asString() instead of name.identifier, so you don't have to check isSpecial first?

Reply:
Right, changed, thanks.